src/storage/index/hnsw_graph.cpp

```diff
@@ -0,0 +1,197 @@
+#include "storage/index/hnsw_graph.h"
+
+#include "storage/store/list_chunk_data.h"
+#include "storage/store/node_table.h"
+#include "storage/store/rel_table.h"
+
+namespace kuzu {
+namespace storage {
+
+InMemEmbeddings::InMemEmbeddings(MemoryManager& mm, EmbeddingTypeInfo typeInfo,
+    common::offset_t numNodes, const common::LogicalType& columnType)
+    : EmbeddingColumn{std::move(typeInfo)} {
+    data = ColumnChunkFactory::createColumnChunkData(mm, columnType.copy(),
+        false /*enableCompression*/, numNodes, ResidencyState::IN_MEMORY, true /*hasNullData*/,
+        false /*initializeToZero*/);
+}
+
+float* InMemEmbeddings::getEmbedding(common::offset_t offset) const {
+    const auto& listChunk = data->cast<ListChunkData>();
+    return &listChunk.getDataColumnChunk()->getData<float>()[offset * typeInfo.dimension];
+}
+
+void InMemEmbeddings::initialize(main::ClientContext* context, NodeTable& table,
+    common::column_id_t columnID) {
+    // Prepare scan state.
+    std::vector<common::column_id_t> columnIDs;
```

Comment:
I hope one day we can hide all these scan-related code behind iterator interfaces so all of these lower-level scan related code can be hidden from the indices, similar to how on-disk graph edge scanners hide the scan logic from GDS algorithms. Many other features we develope will need to scan raw tuples and it looks like we are repeating the code of how to scan data. Ideally, we should just do the following:
```
itr = NodeTable.getColumnScanIterator<int64>(columnID)
while (itr.hasNext()) {
   data->append(itr.getNext());
}
```
And the scanIterator does all the optimizations and boilerplate code that needs to be done. The person developing the Index doesn't need to know how that was done.